kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/Protocol.java

```diff
@@ -55,18 +64,28 @@ public static Protocol fromColumnVector(ColumnVector vector, int rowId) {
 
   private final int minReaderVersion;
   private final int minWriterVersion;
-  private final List<String> readerFeatures;
-  private final List<String> writerFeatures;
+  private final Set<String> readerFeatures;
+  private final Set<String> writerFeatures;
+
+  // These are derived fields from minReaderVersion and minWriterVersion
+  private final boolean supportsReaderFeatures;
+  private final boolean supportsWriterFeatures;
+
+  public Protocol(int minReaderVersion, int minWriterVersion) {
+    this(minReaderVersion, minWriterVersion, null, null);
+  }
 
   public Protocol(
       int minReaderVersion,
       int minWriterVersion,
-      List<String> readerFeatures,
-      List<String> writerFeatures) {
+      Set<String> readerFeatures,
+      Set<String> writerFeatures) {
     this.minReaderVersion = minReaderVersion;
     this.minWriterVersion = minWriterVersion;
     this.readerFeatures = readerFeatures;
```

Comment:
Is there a benefit to using `null`?

I _think_ I'd prefer, at the very least, that callers of `getReaderFeatures` and `getWriterFeatures` do not need to perform null checks. Perhaps storing null internally is fine (and then just expose the empty set)? But perhaps storing an empty Set would be better?

Reply:
+1